Transpile packages on demand, validate all TS projects

spalger opened the pull request:
## Dearest Reviewers 👋 

I've been working on this branch with @mistic and @tylersmalley and we're really confident in these changes. Additionally, this changes code in nearly every package in the repo so we don't plan to wait for reviews to get in before merging this. If you'd like to have a concern addressed, please feel free to leave a review, but assuming that nobody raises a blocker in the next 24 hours we plan to merge this EOD pacific tomorrow, 12/22.

We'll be paying close attention to any issues this causes after merging and work on getting those fixed ASAP. 🚀 

---

The operations team is not confident that we'll have the time to achieve what we originally set out to accomplish by moving to Bazel with the time and resources we have available. We have also bought ourselves some headroom with improvements to babel-register, optimizer caching, and typescript project structure.

In order to make sure we deliver packages as quickly as possible (many teams really want them), with a usable and familiar developer experience, this PR removes Bazel for building packages in favor of using the same JIT transpilation we use for plugins.

Additionally, packages now use `kbn_references` (again, just copying the dx from plugins to packages).

Because of the complex relationships between packages/plugins and in order to prepare ourselves for automatic dependency detection tools we plan to use in the future, this PR also introduces a "TS Project Linter" which will validate that every tsconfig.json file meets a few requirements:

1. the chain of base config files extended by each config includes `tsconfig.base.json` and not `tsconfig.json`
1. the `include` config is used, and not `files`
2. the `exclude` config includes `target/**/*`
3. the `outDir` compiler option is specified as `target/types`
1. none of these compiler options are specified: `declaration`, `declarationMap`, `emitDeclarationOnly`, `skipLibCheck`, `target`, `paths`

4. all references to other packages/plugins use their pkg id, ie:
	
	```js
    // valid
    {
      "kbn_references": ["@kbn/core"]
    }
    // not valid
    {
      "kbn_references": [{ "path": "../../../src/core/tsconfig.json" }]
    }
    ```

5. only packages/plugins which are imported somewhere in the ts code are listed in `kbn_references`

This linter is not only validating all of the tsconfig.json files, but it also will fix these config files to deal with just about any violation that can be produced. Just run `node scripts/ts_project_linter --fix` locally to apply these fixes, or let CI take care of automatically fixing things and pushing the changes to your PR.

> **Example:** [`64e93e5` (#146212)](https://github.com/elastic/kibana/pull/146212/commits/64e93e580679dd55f4fdf19bd01402bc478a1a75) When I merged main into my PR it included a change which removed the `@kbn/core-injected-metadata-browser` package. After resolving the conflicts I missed a few tsconfig files which included references to the now removed package. The TS Project Linter identified that these references were removed from the code and pushed a change to the PR to remove them from the tsconfig.json files.

## No bazel? Does that mean no packages??
Nope! We're still doing packages but we're pretty sure now that we won't be using Bazel to accomplish the 'distributed caching' and 'change-based tasks' portions of the packages project.

This PR actually makes packages much easier to work with and will be followed up with the bundling benefits described by the original packages RFC. Then we'll work on documentation and advocacy for using packages for any and all new code.

We're pretty confident that implementing distributed caching and change-based tasks will be necessary in the future, but because of recent improvements in the repo we think we can live without them for **at least** a year.

## Wait, there are still BUILD.bazel files in the repo
Yes, there are still three webpack bundles which are built by Bazel: the `@kbn/ui-shared-deps-npm` DLL, `@kbn/ui-shared-deps-src` externals, and the `@kbn/monaco` workers. These three webpack bundles are still created during bootstrap and remotely cached using bazel. The next phase of this project is to figure out how to get the package bundling features described in the RFC with the current optimizer, and we expect these bundles to go away then. Until then any package that is used in those three bundles still needs to have a BUILD.bazel file so that they can be referenced by the remaining webpack builds.

spalger commented:
@elasticmachine merge upstream

mistic commented:
I've made some progress on this PR. CI is now green and it is in sync with the latest main. The only thing that catch my eye with this PR and that probably need a bit more work is related to CI workers for Jest unit tests. They doubled with this PR and the CI is consistent around 1h30mins

Commit:
[pkgs] remove pre-building of packages, transpile on demand

Commit:
split up enterprise_search jest config

Commit:
pick a larger default min so that new jest configs which are slow get their own worker

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/ts_project_linter --fix'

Commit:
move pkg resolution to resolver, config is too slow

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

Commit:
Merge branch 'implement/babel-register-packages' of github.com:spalger/kibana into implement/babel-register-packages

Commit:
remove bazel reader

Commit:
add info about weird imports

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

Commit:
move moduleNameMap config to resolver

Commit:
tweat ts project linter after review

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/ts_project_linter --fix'

Commit:
fix APM_AGENT_MOCK path

Commit:
Merge branch 'implement/babel-register-packages' of github.com:spalger/kibana into implement/babel-register-packages

Commit:
restore default resolver for packages using "exports": {} maps

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

Commit:
prioritize checks which auto-commit

elasticmachine commented:
Pinging @elastic/kibana-operations (Team:Operations)

pgayvallet commented on packages/analytics/client/src/analytics_client/analytics_client.ts:
```diff
@@ -171,6 +171,7 @@ export class AnalyticsClient implements IAnalyticsClient {
     this.optInConfig$.next(optInConfigInstance);
   };
 
+  // @ts-expect-error
```
Why do we need that now? Can we at least have a description or a reason? I see a lot of those in our owned changes.

pgayvallet commented on packages/core/apps/core-apps-server-internal/src/bundle_routes/register_bundle_routes.test.ts:
```diff
@@ -56,31 +69,50 @@ describe('registerBundleRoutes', () => {
-      publicPath: '/server-base-path/42/bundles/core/',
-      routePath: '/42/bundles/core/',
-    });
+    expect(registerRouteForBundleMock.mock.calls).toMatchInlineSnapshot(`
```
Why where our proper `toHaveBeenCalledWith` assertions replaced with ugly and long `toMatchInlineSnapshot`?

spalger replied:
Okay, okay, I disagree that these are improper, but my laziness and love for automatically updating snapshots is not something I'll force on the core team 😝

pgayvallet commented on packages/core/http/core-http-router-server-internal/src/request.ts:
```diff
@@ -109,6 +109,7 @@ export class CoreKibanaRequest<
   public readonly rewrittenUrl?: URL;
 
   /** @internal */
+  // @ts-expect-error
   protected readonly [requestSymbol]: Request;
```
Same question, why is this considered a TS error?

pgayvallet commented on packages/core/http/core-http-router-server-internal/src/validator.ts:
```diff
@@ -127,8 +127,10 @@ export class RouteValidator<P = {}, Q = {}, B = {}> {
     namespace?: string
   ): RouteValidationResultType<typeof validationRule> {
     if (isConfigSchema(validationRule)) {
+      // @ts-expect-error
```
Same here.

pgayvallet commented on packages/kbn-es-types/src/search.ts:
```diff
@@ -579,7 +579,8 @@ export type AggregateOf<
 type AggregateOfMap<TAggregationMap extends AggregationMap | undefined, TDocument> = {
   [TAggregationName in keyof TAggregationMap]: Required<TAggregationMap>[TAggregationName] extends AggregationsAggregationContainer
-    ? AggregateOf<TAggregationMap[TAggregationName], TDocument>
+    ? // @ts-expect-error
```
Just recording them for traceability at this point

pgayvallet commented on packages/kbn-health-gateway-server/src/kibana/routes/root.ts:
```diff
@@ -124,6 +124,7 @@ export class RootRoute implements ServerRoute {
     try {
       return await nodeFetch(url, {
         agent: protocol === 'https:' ? this.getAgent() : undefined,
+        // @ts-expect-error
```
+1

pgayvallet commented on packages/kbn-utility-types/src/dot.ts:
```diff
@@ -41,7 +41,8 @@ type DotKey<
   TPrefix extends string
 > = TObject[TKey] extends Array<infer TValueType>
-  ? ToArray<DotObject<TValueType, `${TPrefix}${TKey}.`>>
+  ? // @ts-expect-error
```
+1

pgayvallet commented on src/core/server/mocks.ts:
```diff
@@ -140,7 +140,7 @@ export type CustomRequestHandlerMock<T> = {
 
 const createCustomRequestHandlerContextMock = <T>(contextParts: T): CustomRequestHandlerMock<T> => {
-  const mock = Object.entries(contextParts).reduce(
+  const mock = Object.entries(contextParts as any).reduce(
```
+1

pgayvallet reviewed (commented):
I think I gonna need some better understanding of why the PR had to introduce that many (undocumented) `@ts-expect-error` in our team's owned codebase before being able to approve the changes. 

Feel free to ping me directly, as given it seems you want that merged fast, a sync discussion would probably be best.

stratoula commented:
++ on what mentioned above. I would also like to understand the addition of so many ts-expect-error to our codebase.

watson commented on packages/kbn-safer-lodash-set/BUILD.bazel:
```diff
@@ -1,9 +1,6 @@
 load("@build_bazel_rules_nodejs//:index.bzl", "js_library", "pkg_npm")
```
The other `BUILD.bazel` files has been deleted, but this one hasn't. How come?

spalger replied:
See the last questing in the description, TLDR there are still a few webpack bundles like `@kbn/ui-shared-deps-src` which include some packages, those packages need a BUILD.bazel file for now, but new packages should not be added to these bundles and we're going to start working to remove them next.

watson commented on packages/kbn-user-profile-components/package.json:
```diff
@@ -2,8 +2,5 @@
-  "types": "./target_types/index.d.ts",
   "license": "SSPL-1.0 OR Elastic License 2.0"
-}
+}
```
Did you mean to remove the trailing linebreak?

spalger replied:
That is how our .editorconfig file is setup https://github.com/elastic/kibana/blob/main/.editorconfig#L12-L13

watson commented on packages/kbn-handlebars/package.json:
```diff
@@ -2,8 +2,5 @@
-  "types": "./target_types/index.d.ts",
   "license": "MIT"
-}
+}
```
Did you mean to remove the trailing linebreak?

watson commented on x-pack/test/saved_object_api_integration/common/config.ts:
```diff
@@ -7,7 +7,8 @@
 import path from 'path';
 
-import { REPO_ROOT } from '@kbn/utils';
+// @ts-ignore we have to check types with "allowJs: false" for now, causing this import to fail
```
Wouldn't it be better to use `@ts-expect-error`?

```suggestion
// @ts-expect-error we have to check types with "allowJs: false" for now, causing this import to fail
```

watson commented on x-pack/test/spaces_api_integration/common/config.ts:
```diff
@@ -6,7 +6,8 @@
 
 import path from 'path';
-import { REPO_ROOT } from '@kbn/utils';
+// @ts-ignore we have to check types with "allowJs: false" for now, causing this import to fail
```
Wouldn't it be better to use `@ts-expect-error`?

```suggestion
// @ts-expect-error we have to check types with "allowJs: false" for now, causing this import to fail
```

watson commented on x-pack/plugins/security/server/session_management/session_index.ts:
```diff
@@ -81,6 +81,7 @@ export function getSessionIndexSettings({
   indexName: string;
   aliasName: string;
 }): IndicesCreateRequest {
+  // @ts-expect-error
```
Nit: I personally prefer to add some text to this comment explaining what we're trying to ignore. At a minimum, I just copy the TS error verbatim if it's not too long

watson commented on x-pack/plugins/security/server/authentication/authenticator.ts:
```diff
@@ -274,6 +275,7 @@ export class Authenticator {
           name: '__http__',
           logger: options.loggers.get(HTTPAuthenticationProvider.type),
           urls: {
+            // @ts-expect-error
```
Nit: I personally prefer to add some text to this comment explaining what we're trying to ignore. At a minimum, I just copy the TS error verbatim if it's not too long

watson commented on x-pack/plugins/security/server/authentication/authenticator.ts:
```diff
@@ -258,6 +258,7 @@ export class Authenticator {
               ...providerCommonOptions,
               name,
               logger: options.loggers.get(type, name),
+              // @ts-expect-error
```
Nit: I personally prefer to add some text to this comment explaining what we're trying to ignore. At a minimum, I just copy the TS error verbatim if it's not too long

watson commented on src/plugins/interactive_setup/public/use_form.ts:
```diff
@@ -153,6 +154,7 @@ export function useFormState<Values extends FormValues, Result>({
     const nextErrors = await validateForm(formValues);
     setTouched(mapDeep(formValues, true));
     setSubmitCount(submitCount + 1);
+    // @ts-expect-error
```
Nit: I personally prefer to add some text to this comment explaining what we're trying to ignore. At a minimum, I just copy the TS error verbatim if it's not too long

watson commented on src/plugins/interactive_setup/public/use_form.ts:
```diff
@@ -128,6 +128,7 @@ export function useFormState<Values extends FormValues, Result>({
   const [touched, setTouched] = useState<TouchedFields<Values>>({});
   const [submitCount, setSubmitCount] = useState(0);
 
+  // @ts-expect-error
```
Nit: I personally prefer to add some text to this comment explaining what we're trying to ignore. At a minimum, I just copy the TS error verbatim if it's not too long

watson reviewed (commented):
I've reviewed Kibana Platform Security change only and left a few comments/nits. But feel free to merge as is if you have to

walterra reviewed (approved):
ML changes LGTM. (Saw only one ts-expect-error, just chiming in to say some general clarification on that would be great!)

Thanks for this update, this looks very promising to make working with packages a lot easier!

sebelga reviewed (approved):
LGTM

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

spalger commented:
Apologies @watson and @pgayvallet, I lost track of the fact that I merged https://github.com/elastic/kibana/pull/144469 into this PR at one point, which is where all these errant `@ts-expect-error` directives are coming from. I'm working on pulling it out of this PR and will let you know when that's done.

Commit:
remove ts upgrade

jamster10 reviewed (commented):
On behalf of @elastic/security-threat-hunting-explore, thank you for the cleanup and improvements!  

As noted in other comments it would be ideal to not have any `ts-ignore` / `ts-expect-error` without any explanation.

I found `ts-ignores` in the following files that I think can be swapped to `expect-errors` (thanks for the explanation!):

`x-pack/test/security_solution_endpoint/services/endpoint.ts`

`x-pack/test/security_solution_endpoint/services/endpoint_telemetry.ts`

And the following files have `expect-errors` but no explanation:

`x-pack/test/detection_engine_api_integration/security_and_spaces/group/exception_operators_data_types/keyword_array.ts`

`x-pack/plugins/security_solution/common/utility_types.ts`

`x-pack/plugins/security_solution/public/common/components/page/manage_query.tsx`

`x-pack/plugins/security_solution/public/common/lib/kibana/hooks.ts`

`x-pack/plugins/security_solution/server/lib/detection_engine/routes/utils.ts`

`x-pack/plugins/security_solution/server/lib/risk_score/prebuilt_dev_tool_content/routes/read_prebuilt_dev_tool_content_route.ts`

jamster10 commented on x-pack/plugins/security_solution/common/experimental_features.ts:
```diff
@@ -127,7 +127,7 @@ export const parseExperimentalConfigValue = (configValue: string[]): Experimenta
 };
 
-export const isValidExperimentalValue = (value: string): boolean => {
+export const isValidExperimentalValue = (value: string): value is keyof ExperimentalFeatures => {
```
Thanks for doing this!

Commit:
tsify parts of kbn-es which are breaking tsc

Commit:
remove more errant ts-expect-error directives

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

Commit:
add descriptions to naked `@ts-expect-errors`, remove `@ts-ignore`s

Commit:
fix typo in kbn-es

Commit:
don't force the core team to use snapshots

spalger commented:
@watson @pgayvallet @stratoula @jamster10 I have removed the accidental TS upgrade, and all the `@ts-expect-error/ignore` directives that came with it. I also feel like I've addressed the rest of the feedback as well. Would you mind taking another look please?

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/ts_project_linter --fix'

Commit:
remove "preserveSymlines" setting from tsconfig.base.json, no longer needed

Commit:
Merge branch 'implement/babel-register-packages' of github.com:spalger/kibana into implement/babel-register-packages

spalger commented:
cla/check

shahzad31 reviewed (approved):
Uptime changes LGTM !!

pgayvallet reviewed (approved):
LGTM

Commit:
Merge branch 'main' of github.com:elastic/kibana into implement/babel-register-packages

Commit:
improve ftr types to deal with linting error

watson reviewed (approved):
Kibana Platform Security changes LGTM 👍

Commit:
Merge branch 'main' into implement/babel-register-packages

spalger commented:
Sorry about that @vitaliidm, fixed in https://github.com/elastic/kibana/pull/148407

andrewthad commented:
@Karql has done a pretty good job documenting a possible regression caused by this in https://github.com/elastic/kibana/issues/154325 and I've been able to confirm that this issue is present in both 8.7.0 and 8.7.1. @spalger Are you able to confirm whether or not this PR is the cause?

kbujold commented:
We are experiencing issues with external plugin build and webpack, see [forum post](https://discuss.elastic.co/t/kibana-8-8-8-external-plugin-failure/335240/7) . Could these changes be causing our plugin failure? We do not see this issue when running kibana dev only with a build image. We have no issues with Kibana 8.6.2.

![image](https://github.com/elastic/kibana/assets/26882112/9bbb45e7-69a7-443f-9ac5-5f670156ef87)

brenoandrade-wdrv commented:
This is the [issue](https://github.com/elastic/kibana/issues/159328) where the [PR](https://github.com/elastic/kibana/pull/159630/files) that fixes the above errors resides.
